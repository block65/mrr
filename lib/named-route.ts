import type { Params, ExtractRouteParams } from './types.js';
import { nullOrigin, urlObjectAssign, urlRhs } from './util.js';

type QueryParams = Record<string, string>;

export interface NamedRoute<P extends string, Q extends QueryParams> {
  path: P;
  build: (x?: {
    params?: ExtractRouteParams<P>;
    query?: Q;
    hash?: string;
    origin?: string;
  }) => string;
}

export function interpolate<P extends string>(
  path: P,
  params: ExtractRouteParams<P> | Params = {},
): string {
  return path.replace(/\/:(\w+)[?+*]?/g, (_match, token: keyof typeof params) =>
    params[token] ? `/${params[token]}` : '',
  );
}

export function namedRoute<T extends string, Q extends QueryParams = never>(
  pathname: T,
): NamedRoute<T, Q> {
  return {
    path: pathname,
    build({ params, query, origin, ...rest } = {}) {
      const interpolatedPathname = params
        ? interpolate(pathname, params)
        : pathname;

      const searchParams = new URLSearchParams(query);
      searchParams.sort();

      const newUrl = urlObjectAssign(new URL(origin || nullOrigin), {
        pathname: interpolatedPathname,
        search: searchParams.toString(),
        ...rest,
      });

      return newUrl.origin === nullOrigin.origin
        ? urlRhs(newUrl)
        : newUrl.toString();
    },
  };
}
