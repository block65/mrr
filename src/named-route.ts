export { namedRoute } from '../lib/named-route.js';

export type {
  NamedRoute,
  NamedRouteWithParams,
  NamedRouteWithoutParams,
} from '../lib/named-route.js';
