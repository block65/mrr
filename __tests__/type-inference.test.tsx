import '@testing-library/jest-dom';
import { render } from '@testing-library/react';
import { expect, test } from 'vitest';
import type { ExtractRouteParams } from '../lib/types.js';
import { Route, Router, Routes } from '../src/index.js';
import { namedRoute } from '../src/named-route.js';

const login = namedRoute('/foo/:foo');

const EffCee = (props: ExtractRouteParams<typeof login.path>) => (
  <h1>{props.foo}</h1>
);

test('inferred component function props', async () => {
  const { asFragment } = render(
    <Router>
      <Routes>
        <Route path={login.path}>
          <h1>normal</h1>
        </Route>
        <Route
          path={login.path}
          component={(params) => <h1>{params.foo}</h1>}
        />
        <Route path={login.path} component={EffCee} />
        <Route>Default route with children</Route>
      </Routes>
    </Router>,
  );

  expect(asFragment()).toMatchSnapshot();
});
