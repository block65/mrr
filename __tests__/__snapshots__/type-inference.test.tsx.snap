// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inferred component function props 1`] = `
<DocumentFragment>
  Default route with children
</DocumentFragment>
`;
